Garud H statistics

tomwhite opened the pull request:
This fixes #231.

* This PR depends on #377 for getting a haplotype representation of calls, so that should be merged first. The function there takes genotype calls of shape (variants, samples, ploidy) to (variants, haplotypes), where haplotypes=samples * ploidy.
* If the dataset is not windowed, the variants are assumed to be in a single window. This won't work for large numbers of variants (see discussion about hashing below), so we could choose not to support this and say that the input must be windowed. Would be good to hear thoughts about this @alimanfoo, @jeromekelleher.
* All of the H statistics work by computing statistics on the frequency of occurrences of each haplotype in a cohort. In scikit-allel, the haplotypes (columns of calls) are [hashed](https://github.com/cggh/scikit-allel/blob/5f2f73ff926e105e5a65cacfc28f2928aba70f69/allel/model/ndarray.py#L2507) by calling `hash(x.tobytes())` on the numpy array column `x`. When I tried this on MalariaGEN-scale data using Dask the computation ground to a halt, which I think was due to issues with the GIL. To avoid this, I have written a `hash_columns` function that is Numba-jit compiled, and outperforms the Python hash method by a factor of 5 in a single thread.
* I've used this code in a [notebook](https://nbviewer.jupyter.org/github/tomwhite/shiny-train/blob/sgkit/notebooks/gwss/sgkit_h12.ipynb) to compare with scikit-allel on MalariaGEN. The results for a single cohort and statistic (H12) are concordant (for the first 1000 windows at least), which is promising.
* The H stats notebook that uses scikit-allel has [different window sizes for different cohorts](https://github.com/alimanfoo/shiny-train/blob/master/notebooks/gwss/pop_defs.yml), which is not possible in sgkit easily at the moment (see https://github.com/pystatgen/sgkit/issues/232#issuecomment-722377847 for the same problem with PBS). The most pragmatic way to fix this will probably be to specify the subset of cohorts to calculate the statistic for.

hammer commented:
> which I think was due to issues with the GIL.

cc @ravwojdyla who has debugged GIL contention issues previously.

ravwojdyla commented:
Just dropping some tools I find useful when I suspect GIL issues:
 * [py-spy](https://github.com/benfred/py-spy) which is good at ad-hoc, approximate profiling
 * [gil_load](https://github.com/chrisjbillington/gil_load) which is more accurate but also requires setup

alimanfoo commented:
> To avoid this, I have written a `hash_columns` function that is Numba-jit compiled, and outperforms the Python hash method by a factor of 5 in a single thread.

Amazing!!

alimanfoo commented:
> * If the dataset is not windowed, the variants are assumed to be in a single window. This won't work for large numbers of variants (see discussion about hashing below), so we could choose not to support this and say that the input must be windowed.

Yes, this only makes sense for windowed data, I think it would be fine to require windows.

alimanfoo commented:
Maybe there is a potential problem with hash collisions using dbx33a? https://gist.github.com/91f00a3d327ac07fb23be7cb2e332b4b

tomwhite commented:
> Maybe there is a potential problem with hash collisions using dbx33a?

It looks like the generator is creating duplicates, because it overflows a single-byte unsigned int. It works if you change `u1` to `u4`.

alimanfoo commented:
> > Maybe there is a potential problem with hash collisions using dbx33a?
> 
> It looks like the generator is creating duplicates, because it overflows a single-byte unsigned int. It works if you change `u1` to `u4`.

Beautiful, thanks, sorry for the noise.

jeromekelleher commented on sgkit/stats/popgen.py:
```diff
@@ -682,3 +688,190 @@ def pbs(
+N_GARUD_H_STATS = 4  # H1, H12, H123, H2/H1
+
+
+def _Garud_h(k: ArrayLike) -> ArrayLike:
```
I'm having trouble understanding what ``k`` is here - any chance of a more descriptive name?

Unless k is quite small, isn't ``sorted(collections.Counter(k.tolist()).values(), reverse=True)`` going to be bottleneck?

tomwhite replied:
Changed to `haplotypes`.

I thought that the sorted call would be a bottleneck too, but I haven't seen that during my MalariaGEN testing.

jeromekelleher reviewed (commented):
Looks great @tomwhite.

Given that this is a function that could be applied to very large datasets, I wonder if the reshaping of the genotype data into haplotypes is necessary/worth it. Would we always want to double the memory footprint just to compute the haplotype hashes (which it looks to me is what we're doing)?

We don't have to answer this now, but I thought it was worthwhile asking this question before we add the to-haplotyes function over in #377

tomwhite commented:
Thanks for the review @jeromekelleher. You are right about `to_haplotype_calls` - I don't think we need it. Following your suggestion I have changed the `Garud_h` code to work directly on genotype calls. I did this by changing the hashing function to use `guvectorize` so it can cope with arbitrarily shaped arrays. The [updated notebook](https://nbviewer.jupyter.org/github/tomwhite/shiny-train/blob/sgkit/notebooks/gwss/sgkit_h12.ipynb) is a lot simpler now too (while producing the same output), since there's no need to create a haplotype representation.

I've also removed the non-windowed path following @alimanfoo's suggestion.

jeromekelleher reviewed (approved):
Awesome! This is super cool, thanks @tomwhite !
